### PR TITLE
Native image resources

### PR DESCRIPTION
**Pull request by edward3h**
Here's where I got to with generating the necessary reflection config files to work with Graal native-image.

I included a native image test in the jte-runtime-cp-test-gradle project. I know Gradle stuff and not Maven very much.

I haven't written a Maven version of the test yet :( 

I haven't generated a resource config file for the binary output files. That shouldn't be too hard, just don't have time to do it now.

As mentioned elsewhere I'd like to improve the Gradle plugin but that will be a different branch.

### CODE REVIEW

File: jte-maven-plugin/src/main/java/gg/jte/maven/GeneratorMojo.java [hunk 1]
```diff
@@ -54,6 +54,8 @@
     @Parameter(readonly = true)
     public String packageName = Constants.PACKAGE_NAME_PRECOMPILED;
 
+    @Parameter(defaultValue = "${project.build.directory}/classes")
```
Comment (casid) [inline jte-maven-plugin/src/main/java/gg/jte/maven/GeneratorMojo.java:57]: I think that should be
`@Parameter(defaultValue = "${project.build.directory}/generated-resources/jte")`

to match the `targetDirectory` above.

**Review by casid — COMMENTED** [review]
@edward3h thanks for submitting this PR 👍 

I've added a few review comments. Other than that, this looks very good!

I think it is fine to merge this without maven support in the first place. Maven can still be added later on.

File: jte-gradle-plugin/src/main/java/gg/jte/gradle/JteTaskBase.java [hunk 2]
```diff
@@ -109,4 +110,16 @@ public String getPackageName() {
     public void setPackageName(String packageName) {
         this.packageName = packageName;
     }
+
+    @OutputDirectory
+    @Optional
+    public Path getTargetResourceDirectory()
+    {
```
[the remark this answers is not in the document]
Comment (edward3h) [reply jte-gradle-plugin/src/main/java/gg/jte/gradle/JteTaskBase.java:117]: Will do. My IntelliJ (and my fingers) is set to my employer's style, just need to check format settings.

File: jte/src/main/java/gg/jte/compiler/TemplateCompiler.java [hunk 3]
```diff
@@ -64,13 +64,47 @@ public void cleanAll() {
...
     @Override
-    public List<String> precompileAll() {
+    public List<String> precompileAll()
+    {
         return precompile(codeResolver.resolveAllTemplateNames());
     }
+    
+    private void generateNativeResources(LinkedHashSet<ClassDefinition> classDefinitions)
+    {
+        if (config.resourceDirectory == null)
```
[the remark this answers is not in the document]
Comment (edward3h) [reply jte/src/main/java/gg/jte/compiler/TemplateCompiler.java:79]: That makes sense. Will do.

File: jte-runtime-cp-test-gradle/src/nativeTest/java/gg/jte/NativeTest.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+
+    @Test
+    void exceptionLineNumber1() throws IOException
+    {
+        thenRenderingFailsWithError("exceptionLineNumber1.jte");
+        thenOutputContains("Caused by: java.lang.NullPointerException");
+//        thenOutputContains("Failed to render exceptionLineNumber1.jte, error at exceptionLineNumber1.jte:5");
+        // looks like DebugInfo for line numbers isn't working in native (I didn't dig very deep though)
```
[the remark this answers is not in the document]
Comment (edward3h) [reply jte-runtime-cp-test-gradle/src/nativeTest/java/gg/jte/NativeTest.java:50]: Thanks I'll try that.

File: jte-maven-plugin/src/main/java/gg/jte/maven/GeneratorMojo.java (hunk 1 above)
Comment (edward3h) [reply jte-maven-plugin/src/main/java/gg/jte/maven/GeneratorMojo.java:57]: Ah, this is where I got stuck on how Maven processes resources. For native image to work we want the resources to end up in the same classpath as the classes, and I couldn't see where that would happen. In Gradle this is handled implicitly by the java plugin and processResources task.

Do you know?

**Comment by edward3h** [comment]
I noticed that my native test isn't currently run by the github action - I'll fix that.

File: jte-gradle-plugin/src/main/java/gg/jte/gradle/JteTaskBase.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (edward3h) [reply jte-gradle-plugin/src/main/java/gg/jte/gradle/JteTaskBase.java:117]: By the way, on some of my Github projects I use [Spotless](https://github.com/diffplug/spotless) with a commit hook to enforce style. Is that something you'd like to do?

File: jte-maven-plugin/src/main/java/gg/jte/maven/GeneratorMojo.java (hunk 1 above)
Comment (casid) [reply jte-maven-plugin/src/main/java/gg/jte/maven/GeneratorMojo.java:57]: Oh you're right. It looks like maven doesn't do any resource copying in the phase we're at with the plugin:
https://stackoverflow.com/questions/33166295/how-can-a-plugin-add-its-own-generated-resources

I'd suggest to leave that directory null then (would result in copying the resources next to the classes).

`@Parameter(defaultValue = null)`

If users would want to have that differently, they will need to take care of the copying themselves.

File: jte-runtime-cp-test-gradle/src/nativeTest/java/gg/jte/NativeTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (edward3h) [reply jte-runtime-cp-test-gradle/src/nativeTest/java/gg/jte/NativeTest.java]: Jolly good, that worked.

**Comment by casid** [comment]
@edward3h currently the build fails for Java 1.8 Windows with this error message. I must admit I have no idea what's going on there.

```
> Task :processNativeMainResources NO-SOURCE
FAILURE: Build failed with an exception.
> Task :nativeMainClasses

> Task :jar
* What went wrong:
Execution failed for task ':nativeImage'.
> Task :nativeImage FAILED
> Couldn't find an installation of Windows SDK 7.1 suitable for GraalVM.
```

**Comment by edward3h** [comment]
> @edward3h currently the build fails for Java 1.8 Windows with this error message. I must admit I have no idea what's going on there.
> 

Yeah I noticed, haven't had a chance to look into it yet. What bothers me is that the Java 11 Windows one passed, so it's not that Windows is broken as a whole.

I don't have a Windows machine to test on so I guess I'll look at what images the github actions use.

**Comment by casid** [comment]
It looks like GraalVM 1.8 needs a very ancient SDK for the c++ redistributable. I recently compiled a native binary with GraalVM Java 11 on windows, and even that was a great pain. I'm surprised as well that the Github Action works for 11 on windows!

If there's no easy fix, I'd suggest we just skip the native checks for 1.8 on Windows.

File: jte/src/main/java/gg/jte/compiler/TemplateCompiler.java [hunk 5]
```diff
@@ -64,13 +64,43 @@ public void cleanAll() {
...
         return classDefinitions.stream().map(ClassDefinition::getSourceFileName).collect(Collectors.toList());
     }
 
     @Override
     public List<String> precompileAll() {
         return precompile(codeResolver.resolveAllTemplateNames());
     }
+    
+    private void generateNativeResources(LinkedHashSet<ClassDefinition> classDefinitions) {
+        if (config.resourceDirectory == null) {
```
Comment (casid) [inline jte/src/main/java/gg/jte/compiler/TemplateCompiler.java:77]: Could you add a setting `generateNativeResources` that controls this explicitly?

Besides that the PR looks good to merge 👍

**Review by casid — CHANGES_REQUESTED** [review]
@edward3h I think we should have an explicit setting to control the generation of native resources.

Besides that, looks good to merge!

**Comment by edward3h** [comment]
Haven't looked at the "codecov" checks - should I?

**Comment by casid** [comment]
I think this is due to the fact, that the native code generation is tested by a build script and thus no coverage is collected for that. It's still tested by the integration test, though. I'd tend to not add unit tests just to satisfy codecov.

**Comment by casid** [comment]
Merged 👍

**Pull request merged**